### PR TITLE
Created Warning Message for Duplicate Deps and Test for Issue: 6725

### PR DESCRIPTION
**Pull request by TedYav**
Original Issue: https://github.com/npm/npm/issues/6725

Decisions I made:

- Chose a pretty vanilla warning message. Can change if suggestions are made
- I opted to have it warn if dependencies are same version or different versions

Instead of including in `thenCheckTop` as @iarna  suggested, I opted to factor the duplicate checking to a separate function. I thought this kept things a little cleaner.

Wrote test as well. I could unit test the function in particular but I think this test verifies that it's properly created the warning.

### CODE REVIEW

**Comment by TedYav** [comment]
Edit: fixed bug with `idealTree.package.dependencies` or `idealTree.package.devDependencies` being undefined for test `normalize-package-explode.js`

**Comment by legodude17** [comment]
That does look like a config error...

**Comment by iarna** [comment]
Those things quoted aren't actually errors. The first is the summary of the results and the second is... well, no problem at all.

But no, appveyor has never been passing. IIRC, the git tests don't pass on it (even though they pass on a real windows machine).

File: AUTHORS [hunk 1]
```diff
@@ -451,3 +451,4 @@ Dmitry Litvinchenko <karaliti@gmail.com>
 chocolateboy <chocolate@cpan.org>
 Henry Zhu <hi@henryzoo.com>
 Nate Goldman <ungoldman@gmail.com>
+Ted Yavuzkurt <hello@TedY.io>
```
Comment (iarna) [inline AUTHORS]: AUTHORS is machine generated (when we do releases). To use a name/email address other than the one on your git commits we add an entry to `.mailcap`.

File: lib/install/validate-tree.js [hunk 2]
```diff
@@ -73,5 +74,31 @@ function thenCheckTop (idealTree, next) {
...
+  next()
+}
+
+// check if dependencies are duplicated between dependencies and
+// dev-dependencies. Issue warning if so.
+function thenCheckDuplicateDeps (idealTree, next) {
+  var shorterDepsList = idealTree.package.devDependencies
+  var longerDepsList = idealTree.package.dependencies
+  // might be undefined -- skip if so
+  if (!(idealTree.package.devDependencies === undefined || idealTree.package.dependencies === undefined)) {
```
Comment (iarna) [inline lib/install/validate-tree.js]: I would prefer to just set defaults for the deplists and skip this check, eg:

`var shorterDepsList = idealTree.package.devDependencies || {}`

File: lib/install/validate-tree.js [hunk 3]
```diff
@@ -73,5 +74,31 @@ function thenCheckTop (idealTree, next) {
...
+    // speed optimization --> iterate through shorter dependencies list
```
Comment (iarna) [inline lib/install/validate-tree.js]: The longest possible list is going to have a couple hundred items here and this code is called once, so I don't think optimization is justified.

**Review by iarna — APPROVED** [review]
Thanks for putting this together! I'll be making a few minor tweaks (as discussed in the review and merging it.

File: AUTHORS (hunk 1 above)
Comment (TedYav) [inline AUTHORS]: Ah, okay thanks :) Wasn't sure!

File: lib/install/validate-tree.js (hunk 2 above)
Comment (TedYav) [inline lib/install/validate-tree.js]: Good idea—makes it simpler.

**Comment by TedYav** [comment]
Okay, sounds good, will keep that in mind in future. Output I pasted from AppVeyor was just showing that I thought failing tests might be related to CouchDB not being running (thus causing them to time out or something). Didn't dig in too closely.

**Comment by iarna** [comment]
Thank you again for this. It's released in v4.4.0!

**Pull request closed without merging**